Refactoring convert array

jinsooihm opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Shape check moved into `convert_array` function.
`squeeze` and `atleast_2d` is not implemented yet.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Related to #199.
Resolves #268.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Few tests have been added and it passes all the existing tests.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

jinsooihm committed:
Added shape check into convert_array

jinsooihm committed:
Removed error message parameter

jinsooihm committed:
Removed shape checking

jinsooihm committed:
Added documentation

vyasr reviewed (changes requested):
After discussion, we'd like to merge the shape and dimension checks into one check. That should simplify this further.

jinsooihm committed:
Removed dimension parameter

jinsooihm committed:
Fixed some local issue

Build trash was doing something weird locally, and would not really use the new common.py

jinsooihm commented:
Done! 
One side note:
I was in the branch `feature/property` and there `common` is a `pyx` file and thus produces a `common.cpython-36m-darwin.so` file in the directory `freud`.
When I checked out this branch, that `common.cpython-36m-darwin.so` file still exists, and importing `common` will not import `common.py`, but instead will import the implementation in `common.cpython-36m-darwin.so` file. (This can be checked from just opening a python interpreter and try printing `freud.common.__file__`.)
It really messed things up on my local machine because all the changes to `common.py` does not get reflected, and even running `python setup.py clean` does not clean it.

vyasr commented:
Thanks for the heads up. Makes sense, that annoyance probably won't go away until we merge that branch, so good to keep it in mind.

vyasr commented on freud/box.pyx:
```diff
@@ -391,7 +391,8 @@ cdef class Box:
 
         vecs = freud.common.convert_array(vecs)
-        imgs = freud.common.convert_array(imgs, vecs.ndim, dtype=np.int32)
+        imgs = freud.common.convert_array(imgs, (None, ) * vecs.ndim,
```
Can you just make this check `vecs.shape` and then remove the extra check above?

vyasr committed:
Minor changes

vyasr reviewed (approved):
Everything looks good! There were a couple of small things I wanted changed, but they were were all tiny changes so it made more sense for me to just make them myself.

vyasr committed:
Fix style

vyasr committed:
Fix wraparound

vyasr committed:
Fix error type

bdice commented on freud/cluster.pyx:
```diff
@@ -131,10 +128,8 @@ cdef class Cluster:
             keys((:math:`N_{particles}`) :class:`numpy.ndarray`):
```
Is this missing a space after `keys`?

bdice commented on freud/cluster.pyx:
```diff
@@ -245,16 +240,9 @@ cdef class ClusterProperties:
-            raise RuntimeError(
-                ('cluster_idx must be a 1D array of matching length/number'
-                    'of particles to points'))
+            cluster_idx, dtype=np.uint32, shape=(points.shape[0],))
```
Switch the order of arguments.

bdice commented on freud/common.py:
```diff
@@ -10,30 +10,40 @@
+        shape: (tuple of int and :code:`None`): Expected shape of the array.
+            Only the dimensions that are not :code:`None` are checked.
+            (Default value = :code:'None').
         dtype: code:`dtype` to convert the array to if :code:`array.dtype`
```
Missing colon before `code:` ➡️ `:code:`

bdice commented on freud/environment.pyx:
```diff
@@ -435,13 +422,8 @@ cdef class LocalDescriptors:
-                raise ValueError(
-                    "orientations must have the same size as points_ref")
+            orientations = freud.common.convert_array(orientations,
+                shape=(points_ref.shape[0], 4)) # noqa
```
Not sure why this is `# noqa` - is there a `flake8` complaint? I like to mark `# noqa: E501` or whatever the `flake8` rule is that it breaks.

bdice commented on freud/locality.pyx:
```diff
@@ -284,7 +284,8 @@ cdef class NeighborQuery:
         # Can't use this function with old-style NeighborQuery objects
-        points = freud.common.convert_array(np.atleast_2d(points), 2)
+        points = freud.common.convert_array(np.atleast_2d(points),
+                                            (None, None))
```
Why not `shape=(None, 3)`?

bdice commented on freud/locality.pyx:
```diff
@@ -459,16 +458,16 @@ cdef class NeighborList:
-        index_i = freud.common.convert_array(index_i, 1, dtype=np.uint64)
-        index_j = freud.common.convert_array(index_j, 1, dtype=np.uint64)
+        index_i = freud.common.convert_array(index_i, (None,), dtype=np.uint64)
+        index_j = freud.common.convert_array(index_j, (None,), dtype=np.uint64)
```
We could get rid of the shape check below by enforcing `index_j` has the same shape as `index_i`.

bdice commented on freud/locality.pyx:
```diff
@@ -459,16 +458,16 @@ cdef class NeighborList:
             weights = np.ones(index_i.shape, dtype=np.float32)
         else:
-            weights = freud.common.convert_array(weights, 1)
+            weights = freud.common.convert_array(weights, (None, ))
```
Check to make sure `weights` is the same shape as `index_i`.

bdice commented on freud/locality.pyx:
```diff
@@ -812,7 +807,8 @@ cdef class AABBQuery(NeighborQuery):
             self._box = freud.common.convert_box(box)
-            self.points = freud.common.convert_array(points, 2).copy()
+            self.points = freud.common.convert_array(
+                points, (None, None)).copy()
```
As above, can we check for shape `(None, 3)`?

bdice commented on freud/locality.pyx:
```diff
@@ -827,7 +823,8 @@ cdef class AABBQuery(NeighborQuery):
         # Can't use this function with old-style NeighborQuery objects
-        points = freud.common.convert_array(np.atleast_2d(points), 2)
+        points = freud.common.convert_array(
+            np.atleast_2d(points), (None, None))
```
`(None, 3)`?

bdice commented on freud/msd.pyx:
```diff
@@ -142,24 +142,10 @@ cdef class MSD:
-            if not positions.shape == images.shape:
-                raise TypeError(
-                    'The positions and images must have the same shape')
+            images = freud.common.convert_array(images, dtype=np.int32,
```
Swap dtype and shape.

bdice commented on freud/order.pyx:
```diff
@@ -233,7 +231,7 @@ cdef class NematicOrderParameter:
         cdef vec3[float] l_u = vec3[float](u[0], u[1], u[2])
         self.thisptr = new freud._order.NematicOrderParameter(l_u)
-        self.u = freud.common.convert_array(u, 1)
+        self.u = freud.common.convert_array(u, (None, ))
```
Should be shape `(3,)`

bdice commented:
We're pretty inconsistent throughout about whether we define the shape as a keyword argument or just a plain argument, like `freud.common.convert_array(arr, shape=(None, 3))` vs. `freud.common.convert_array(arr, (None, 3))`. I like defining it as a keyword argument for clarity, but that's a stylistic preference. Consistency would be good either way.

bdice commented:
Also check for other instances of the `(None, None)` pattern that I mentioned - I stopped leaving comments after the first few. It doesn't seem right to me (should be `(None, 3)` if I understand correctly) but it was repeated a lot so I may have misinterpreted something.

bdice reviewed (changes requested):
See comments.

vyasr committed:
Address PR comments

vyasr committed:
Standardize usage of keyword args for convert_array

vyasr commented:
The majority of cases were using keyword args, so I switched to doing that everywhere.

vyasr committed:
Fix flake8 again

bdice commented on freud/environment.pyx:
```diff
@@ -149,14 +149,14 @@ cdef class BondOrder:
             box (:class:`freud.box.Box`):
                 Simulation box.
-            ref_points ((:math:`N_{particles}`, 3) :class:`numpy.ndarray`):
+            ref_points ((:math:`N_{ref_points}`, 3) :class:`numpy.ndarray`):
```
I think the docs will render this wrong unless you escape the underscore. Please check.

bdice reviewed (approved):
One doc fix - then good to merge. 👍

bdice committed:
Merge branch 'master' into refactor/convert_array2

bdice committed:
Fix underscore in ref_points.
